src/tsdb.c

```diff
@@ -224,85 +225,128 @@ int SeriesAddSample(Series *series, api_timestamp_t timestamp, double value) {
+
     void *dictResult = (void *)TRUE;
-    while (dictResult) {
-        if (funcs->GetLastTimestamp(iter->currentChunk) < minTimestamp) {
-            dictResult = RedisModule_DictNextC(iter->dictIter, NULL, (void*)&iter->currentChunk);
-        } else {
-            iter->chunkIterator = funcs->NewChunkIterator(iter->currentChunk);
-            iter->minTimestamp = minTimestamp;
-            iter->maxTimestamp = maxTimestamp;
-            return REDISMODULE_OK;
-        }
+    if (iter->reverse == NO_OPT) {
+        iter->dictIter = RedisModule_DictIteratorStartC(series->chunks, "<=", &rax_key, sizeof(rax_key));
+        RedisModule_DictNextC(iter->dictIter, NULL, (void*)&iter->currentChunk);
+        while (dictResult) {
+            if (funcs->GetLastTimestamp(iter->currentChunk) < iter->minTimestamp) {
+                dictResult = RedisModule_DictNextC(iter->dictIter, NULL, (void*)&iter->currentChunk);
+            } else break;
+        }        
+    } else {    
+        iter->dictIter = RedisModule_DictIteratorStartC(series->chunks, "$", &rax_key, sizeof(rax_key));
+        RedisModule_DictPrevC(iter->dictIter, NULL, (void*)&iter->currentChunk);
+        while (dictResult) {
+            if (funcs->GetFirstTimestamp(iter->currentChunk) > iter->maxTimestamp) {
+                dictResult = RedisModule_DictPrevC(iter->dictIter, NULL, (void*)&iter->currentChunk);
+            } else break;
+        }      
     }
-    SeriesIteratorClose(iter);
-    return REDISMODULE_ERR;
+    
+    if (dictResult == NULL) {
+        return REDISMODULE_ERR;
+    }
+    iter->chunkIterator = funcs->NewChunkIterator(iter->currentChunk, iter->reverse);
+    return REDISMODULE_OK;
 }
 
+// Fill the first sample
 ChunkResult SeriesIteratorGetFirst(SeriesIterator *iterator, Sample *sample) {
+    ChunkResult res;
     ChunkFuncs *funcs = iterator->series->funcs;
-    ChunkResult res = funcs->ChunkIteratorGetNext(iterator->chunkIterator, sample);
 
     if (funcs->GetNumOfSample(iterator->currentChunk) == 0) {
         return CR_END;
     }
 
-    if (sample->timestamp < iterator->minTimestamp) {
-        while (res == CR_OK) {     
-            // Skip through initial samples
-            res = funcs->ChunkIteratorGetNext(iterator->chunkIterator, sample);
-            if (sample->timestamp >= iterator->minTimestamp) { 
-                break;
+    if (iterator->reverse == NO_OPT) {
+        res = funcs->ChunkIteratorGetNext(iterator->chunkIterator, sample);
+        if (sample->timestamp < iterator->minTimestamp) {
+            while (res == CR_OK) {     
+                // Skip through initial samples
+                res = funcs->ChunkIteratorGetNext(iterator->chunkIterator, sample);
+                if (sample->timestamp >= iterator->minTimestamp) { 
+                    break;
+                }
+            }
+            return res; // shouldn't happen
+        }
+    } else {
+        res = funcs->ChunkIteratorGetPrev(iterator->chunkIterator, sample);
```

Comment:
@ashtul this is where `SeriesIteratorGetFirst` is very confusing. this is not the *first* sample in the series, its the *first* sample in the iterator.
Since the name mentions both, it can be misleading.

Reply:
Is this a naming question or design question?
This function gets the first sample for a series-iterator. I'm open for naming suggestions